src/WMBot/Commands/Admin.cs

```diff
@@ -109,6 +110,89 @@ private static void CommandList(CommandParams parameters)
                 IRC.DeliverMessage("I know: " + commands, parameters.SourceUser);
         }
 
+        private static void Restrict(CommandParams parameters)
+        {
+            if (string.IsNullOrEmpty(parameters.Parameters))
+            {
+                IRC.DeliverMessage("Please specify restrict action", parameters.SourceChannel);
+                return;
+            }
+
+            List<string> pm = new List<string>(parameters.Parameters.Trim().Split(' '));
+            switch (pm[0])
+            {
+                case "add":
+                    if (pm.Count == 1)
+                    {
+                        IRC.DeliverMessage("This command requires channel name to be specified", parameters.SourceChannel.Name);
+                        break;
+                    }
+
+                    string channel_name = pm[1];
+                    bool allow_adding_by_admin = pm.Count >= 3 && pm[2] == "admin";
+
+                    if (!Core.ValidFile(channel_name) || !channel_name.StartsWith("#"))
+                    {
+                        IRC.DeliverMessage(messages.Localize("InvalidName", parameters.SourceChannel.Language, new List<string> { channel_name }), parameters.SourceChannel);
+                        return;
+                    }
+
+                    parameters.SourceChannel.PrimaryInstance.Network.Transfer("PART " + channel_name);
```

Comment:
I am not sure what is going on here, are we parting the channel because bot was in it and we just decided to restrict access to that channel? If yes, where do we check if bot even was in the channel in first place? :)

also this wouldn't work 100% because you are calling it on primary instance of SourceChannel, but target channel may not be on same instance as source channel (source channel is where the command is coming from).

I understand that concept of instances is confusing, but basically each instance is a separate bot (as in IRC session), all these sessions share the same core process. There are multiple reasons why we use multiple bot instances, but main are performance and network balancing.

Reply:
Yes, you are right. We are parting the channel because the bot was there and we restricted access to that channel.

I will change this so we part the channel only if the bot was there.
I will also make this code to choose correct bot instance when parting.

Reply:
Before you start working on it, check my last comment on bottom and reconsider if it wouldn't be easier to make simple bad-channel list instead of modifying the internal logic of how channels are being handled. Right now every existing instance of Channel class that is also in Configuration.Channels are considered to be active, live channels that the bot is currently in. On start of bot, it will attempt to automatically join all of them, and it's even possible that there is (if not I had plans to implement it), some code that periodically checks if bot really is in all these channels and if not, it would try to rejoin it.

If you really want to go this way, then this logic needs to change in code as well and bot needs to treat these "bad" channels in different way than normal channels

Reply:
Hm. I see what you mean. You are right. It is easier and better to have a bad-channel list. The only question now is what is the best way to store that channel could be unblocked not just by root, but also by admin. Shall we have a separate list for that purpose?